### DISCUSSION TITLE
Using function like things in templates

### DISCUSSION THREAD

**Discussion by danielo515** [twpayne/chezmoi#2947] [Show and tell]
Today I had a conversation with chat-gpt to update one of my chezmoi templates. I'm not good at templating and I wanted it to do it fast. I asked it to add some comments to the generated file so I can see where each section is coming from in the generated file. Because I had some conditionals, it duplicated the comment in several places. When I asked it if there is a way to avoid the duplication, to my surprise it suggested to use the define directive and created a simple include block with comments. I searched the docs and I didn't find anything mentoning this, but it works, so I want to share it here:

```tpl
{{ template "variables" . }}
{{ template "path" . }}
{{- $chezmoiOs := .chezmoi.os -}}
{{- $zshFiles := glob (joinPath .chezmoi.workingTree "zsh" "*.zsh") -}}

{{- define "includeWithComments" -}}
# ### START OF {{ .filename }} ###
{{ include .filename }}
# ### END OF {{ .filename }} ###
{{ end }}

{{- range $zshFile := $zshFiles -}}
{{- if hasSuffix "osx.zsh" $zshFile -}}
  {{- if eq $chezmoiOs "darwin" -}}
    {{ template "includeWithComments" (dict "filename" $zshFile) }}
  {{- end -}}
{{- else -}}
  {{ template "includeWithComments" (dict "filename" $zshFile) }}
{{- end -}}
{{- end -}}
```

**Comment by bradenhilton** [comment]
[`text/template` documentation](https://pkg.go.dev/text/template#hdr-Nested_template_definitions) for `define`.

**Reply by danielo515** [reply]
That is very useful for go programmers, but I don't have any interest in golang. I think this kind of recipes should be mentioned for those people (like me) who are not go programmers

**Reply by twpayne** [reply]
If you're not prepared to read the documentation then there's not much that can be done to help you.

**Reply by danielo515** [reply]
I'm prepared to read the documentation of the tools I use, not the documentation of the whole dependency chain they use. IMO is not reasonable to expect users of a tool to be go programmers just because the tool is written in go. I get that go programmers are very passionate about their language, but I have no interest on it, and I don't want to be a go programmer.
I just saw what I considered a nice way of saving some boilerplate and I wanted to share it with the community for other users that, like me, are not go programmers.
This specific recipe I think is common and it is not even mentioned in chezmoi docs as something possible. Please let me know how you expect a user like me to get to this go template feature without reading the entire go templates docs. If I don't even know this feature exist, how would I try to read its docs? Also, for people used to both chezmoi and go, may be very easy to know what things are special chezmoi defined functions, and what are go primitives, but for me it is very confusing.

**Reply by twpayne** [reply]
Did you even look at the link? It describes the `text/template` language that chezmoi uses. The chezmoi user guide also has a complete [section on templating](https://www.chezmoi.io/user-guide/templating/).

chezmoi made open source and free of charge by its maintainers, who volunteer their time for free. Please read [this article](https://tylercipriani.com/blog/2023/01/15/entitlement-burnout-and-toxicity/) before continuing this discussion.
